src/util/compilationDatabase.ml

```diff
@@ -68,17 +68,20 @@ let load_and_preprocess ~all_cppflags filename =
             preprocess_command
         | None, Some arguments ->
           let arguments = List.map reroot arguments in
-          begin match List.findi (fun i e -> e = "-o") arguments with
-            | (o_i, _) ->
-              begin match List.split_at o_i arguments with
-                | (arguments_program :: arguments_init, _ :: o_file :: arguments_tl) ->
-                  let preprocess_arguments = all_cppflags @ "-E" :: "-MMD" :: "-MT" :: file :: arguments_init @ "-o" :: preprocessed_file :: arguments_tl in
-                  Filename.quote_command arguments_program preprocess_arguments
-                | _ ->
-                  failwith "CompilationDatabase.preprocess: no -o argument value found for " ^ file
-              end
-            | exception Not_found ->
-              failwith "CompilationDatabase.preprocess: no -o argument found for " ^ file
+          let preprocess_arguments =
+            let suf =
+              begin match List.findi (fun i e -> e = "-o") arguments with
+                | (o_i, _) ->
+                  begin match List.drop o_i arguments with
+                    | (_ :: o_file :: arguments_tl) -> preprocessed_file :: arguments_tl
+                    | _ -> failwith ("CompilationDatabase.preprocess: no argument found for -o option for " ^ file)
+                  end
+                | exception Not_found -> [preprocessed_file]
+              end in
+            all_cppflags @ "-E" :: "-MMD" :: "-MT" :: file :: arguments @ "-o" :: suf in
```

Comment:
Or umm, this `arguments` duplicates some of `-o` exists, then `arguments_tl` are included twice (once in `arguments` and once in `suf`). Moreover, `arguments` also contains the old `-o` argument in addition to the new one. Maybe this happens to work because one comes after the other, but this is fragile.

The old approach should be preferred that actually removes the old `-o` argument and doesn't duplicate or miss anything.

Reply:
Thanks for the note, I must have missed that when refactoring.